### PR TITLE
collect IP address from containers in awsvpc mode

### PR DESCRIPTION
**Pull request by hkaj**
### What does this PR do?

collect IP address from containers in awsvpc mode

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/util/docker/containers.go [hunk 1]
```diff
@@ -190,10 +191,28 @@ func parseContainerHealth(status string) string {
...
 // parseContainerNetworkAddresses converts docker ports
 // and network settings into a list of NetworkAddress
-func parseContainerNetworkAddresses(ports []types.Port, netSettings *types.SummaryNetworkSettings, container string) []containers.NetworkAddress {
+func (d *DockerUtil) parseContainerNetworkAddresses(cID string, ports []types.Port, netSettings *types.SummaryNetworkSettings, container string) []containers.NetworkAddress {
 	addrList := []containers.NetworkAddress{}
 	tempAddrList := []containers.NetworkAddress{}
 	if netSettings == nil || len(netSettings.Networks) == 0 {
+		networkMode, err := GetContainerNetworkMode(cID)
+		if err != nil {
+			log.Debugf("failed to get network mode for container %s. Network info will be missing. Error: %s", cID, err)
```
Comment (ahmed-mez) [inline pkg/util/docker/containers.go]: nit
```suggestion
			log.Debugf("Failed to get network mode for container %s. Network info will be missing. Error: %s", cID, err)
```

File: pkg/util/docker/containers.go [hunk 2]
```diff
@@ -190,10 +191,28 @@ func parseContainerHealth(status string) string {
...
+			return addrList
+		}
+		if networkMode == containers.AwsvpcNetworkMode {
+			ecsContainerMetadataUrl, err := d.getECSMetadataURL(cID)
+			if err != nil {
+				log.Debugf("failed to get the ECS container metadata URI for container %s. Network info will be missing. Error: %s", cID, err)
```
Comment (ahmed-mez) [inline pkg/util/docker/containers.go]: same

File: pkg/util/docker/containers.go [hunk 3]
```diff
@@ -190,10 +191,28 @@ func parseContainerHealth(status string) string {
 
 // parseContainerNetworkAddresses converts docker ports
 // and network settings into a list of NetworkAddress
-func parseContainerNetworkAddresses(ports []types.Port, netSettings *types.SummaryNetworkSettings, container string) []containers.NetworkAddress {
+func (d *DockerUtil) parseContainerNetworkAddresses(cID string, ports []types.Port, netSettings *types.SummaryNetworkSettings, container string) []containers.NetworkAddress {
 	addrList := []containers.NetworkAddress{}
 	tempAddrList := []containers.NetworkAddress{}
 	if netSettings == nil || len(netSettings.Networks) == 0 {
```
Comment (ahmed-mez) [inline pkg/util/docker/containers.go]: could we add a comment here to explain that `len(netSettings.Networks) == 0` might correspond to a `awsvpc` mode?

File: pkg/util/docker/containers.go [hunk 4]
```diff
@@ -190,10 +191,28 @@ func parseContainerHealth(status string) string {
...
+		}
+		if networkMode == containers.AwsvpcNetworkMode {
+			ecsContainerMetadataUrl, err := d.getECSMetadataURL(cID)
+			if err != nil {
+				log.Debugf("failed to get the ECS container metadata URI for container %s. Network info will be missing. Error: %s", cID, err)
+				return addrList
+			}
+			res, err := metadata.GetContainerNetworkAddresses(cID, ecsContainerMetadataUrl)
+			if err != nil {
+				log.Errorf("failed to get container network addresses: %s", err)
```
Comment (ahmed-mez) [inline pkg/util/docker/containers.go]: same

File: pkg/util/docker/network.go [hunk 5]
```diff
@@ -143,99 +138,27 @@ func resolveDockerNetworks(containerNetworks map[string][]dockerNetwork) {
...
-// enp0s3  00000000     0202000A  0003   0       0    0       00000000  0    0       0
-// enp0s3  0002000A     00000000  0001   0       0    0       00FFFFFF  0    0       0
-//
-// The returned value would be ["enp0s3","00000000","0202000A","0003","0","0","0","00000000","0","0","0"]
-//
-func defaultGatewayFields() ([]string, error) {
-	procRoot := config.Datadog.GetString("proc_root")
-	netRouteFile := filepath.Join(procRoot, "net", "route")
-	f, err := os.Open(netRouteFile)
+// GetAgentContainerNetworkMode provides the network mode of the Agent container
```
Comment (ahmed-mez) [inline pkg/util/docker/network.go]: could you add these comments back? 
```
// To get this info in an optimal way, consider calling util.GetAgentNetworkMode	func GetContainerNetworkMode(cid string) (string, error) {
// instead to benefit from the cache
```

**Review by ahmed-mez — DISMISSED** [review]
few comments but LGTM otherwise, also apparently some of the refactored modules are missing the `+docker` build tag that's why the CI is failing on windows.

File: pkg/util/containers/network.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// DefaultGateway returns the default Docker gateway.
+func DefaultGateway() (net.IP, error) {
```
Comment (leeavital) [inline pkg/util/containers/network.go]: I think you might need to merge master in, this code has already merged

File: pkg/util/docker/containers.go [hunk 7]
```diff
@@ -190,10 +191,28 @@ func parseContainerHealth(status string) string {
 
 // parseContainerNetworkAddresses converts docker ports
 // and network settings into a list of NetworkAddress
-func parseContainerNetworkAddresses(ports []types.Port, netSettings *types.SummaryNetworkSettings, container string) []containers.NetworkAddress {
+func (d *DockerUtil) parseContainerNetworkAddresses(cID string, ports []types.Port, netSettings *types.SummaryNetworkSettings, container string) []containers.NetworkAddress {
```
Comment (leeavital) [inline pkg/util/docker/containers.go]: I have the bridge mode/host mode handling happening further up in DockerUtil.List() in https://github.com/DataDog/datadog-agent/pull/4418  -- should that move here as well?

File: pkg/util/containers/network.go (hunk 6 above)
Comment (hkaj) [reply pkg/util/containers/network.go]: i'm actually moving it from util/docker to util/containers, because otherwise we get an import cycle

**Review by leeavital — DISMISSED** [review]

File: pkg/util/docker/containers.go (hunk 7 above)
Comment (hkaj) [reply pkg/util/docker/containers.go]: yep, can do 👍

File: pkg/util/docker/network.go (hunk 5 above)
Comment (hkaj) [reply pkg/util/docker/network.go]: ah yeah sorry

File: pkg/util/docker/containers.go [hunk 8]
```diff
@@ -250,6 +280,24 @@ func isExposed(port types.Port) bool {
...
 
+// getECSMetadataURL inspects a given container ID and returns its ECS container metadata URI
+// if found in its environment. It returns an empty string and an error on failure.
+func (d *DockerUtil) getECSMetadataURL(cID string) (string, error) {
+	i, err := d.Inspect(cID, false)
+	if err != nil {
+		return "", err
+	}
+	for _, e := range i.Config.Env {
+		split := strings.Split(e, "=")
```
Comment (leeavital) [inline pkg/util/docker/containers.go]: Nit: you could use strings.HasPrefix to avoid an allocation for every environment variable.

**Review by leeavital — DISMISSED** [review]

File: pkg/util/docker/containers.go (hunk 8 above)
Comment (hkaj) [reply pkg/util/docker/containers.go]: that's a good point!

**Pull request merged**